cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -293,76 +315,172 @@ func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
 }
 
 type PaginationOptions struct {
-	Cursor *events.Cursor `json:"cursor,omitempty"`
-	Limit  uint           `json:"limit,omitempty"`
+	Cursor *db.Cursor `json:"cursor,omitempty"`
+	Limit  uint       `json:"limit,omitempty"`
 }
 
 type GetEventsResponse struct {
 	Events       []EventInfo `json:"events"`
 	LatestLedger uint32      `json:"latestLedger"`
 }
 
-type eventScanner interface {
-	Scan(eventRange events.Range, f events.ScanFunction) (uint32, error)
-}
-
 type eventsRPCHandler struct {
-	scanner      eventScanner
+	dbReader     db.EventReader
 	maxLimit     uint
 	defaultLimit uint
+	logger       *log.Entry
+	ledgerReader db.LedgerReader
 }
 
-func (h eventsRPCHandler) getEvents(request GetEventsRequest) (GetEventsResponse, error) {
+func combineContractIDs(filters []EventFilter) ([][]byte, error) {
+	contractIDSet := set.NewSet[string](maxFiltersLimit * maxContractIDsLimit)
+
+	for _, filter := range filters {
+		for _, contractID := range filter.ContractIDs {
+			contractIDSet.Add(contractID)
+		}
+	}
+
+	contractIDs := make([][]byte, 0, len(contractIDSet))
+	for contractID := range contractIDSet {
+		id, err := strkey.Decode(strkey.VersionByteContract, contractID)
+		if err != nil {
+			return nil, fmt.Errorf("invalid contract ID %v: ", contractID)
+		}
+
+		contractIDs = append(contractIDs, id)
+	}
+	return contractIDs, nil
+}
+
+func combineEventTypes(filters []EventFilter) []int {
+	eventTypes := make(map[int]bool)
+	for _, filter := range filters {
+		for _, eventType := range filter.EventType.Keys() {
+			eventTypeXDR := getEventTypeXDRFromEventType()[eventType]
+			eventTypes[int(eventTypeXDR)] = true
+		}
+	}
+	uniqueEventTypes := make([]int, 0, len(eventTypes))
+	for eventType := range eventTypes {
+		uniqueEventTypes = append(uniqueEventTypes, eventType)
+	}
+	return uniqueEventTypes
+}
+
+func combineTopics(filters []EventFilter) ([][][]byte, error) {
+	encodedTopicsList := make([][][]byte, db.MaxTopicCount)
```

Comment:
I think you want `0, db.MaxTopicCount`, right? Otherwise you have empty bytes and then the `append` adds to the end of that? :thinking:

Reply:
No no if I do that it will fail at `encodedTopicsList[i] = append(encodedTopicsList[i], encodedTopic)` because it is 0 length but on the other hand I have created 4 nil slices which I can easily access using `encodedTopicsList[i]`